asllib/aslspec/PP.ml

```diff
@@ -100,6 +100,87 @@ let pp_variants_with_attributes fmt variants =
 
 let pp_variants fmt variants = pp_sep_list ~sep:" | " pp_type_term fmt variants
 
+let expr_operator_to_string =
+  let open Rule in
+  function
+  | Operator_Assign -> tok_str COLON_EQ
+  | Operator_Equal -> tok_str EQ
+  | Operator_Iff -> tok_str IFF
+  | Operator_List -> tok_str LIST
+  | Operator_Set -> tok_str SET
+  | Operator_Size -> tok_str SIZE
+  | Operator_Some -> tok_str SOME
+  | Operator_Union -> tok_str UNION
+  | Operator_UnionList -> tok_str UNION_LIST
+
+let pp_output_opt fmt is_output = if is_output then fprintf fmt "--@," else ()
+
+let rec pp_expr fmt =
+  let open Rule in
+  function
+  | Var name -> pp_print_string fmt name
+  | Application { applicator = ExprOperator op; args = [ lhs; rhs ] }
+    when is_infix_operator op ->
+      fprintf fmt "%a %s %a" pp_expr lhs
+        (expr_operator_to_string op)
+        pp_expr rhs
+  | Application { applicator; args } ->
+      fprintf fmt "%a(%a)" pp_application_lhs applicator (pp_comma_list pp_expr)
+        args
+  | FieldAccess path -> fprintf fmt "%s" (String.concat "." path)
+  | ListIndex { var; index } -> fprintf fmt "%s[%s]" var index
+  | Record { label; fields } ->
+      fprintf fmt "%s[%a]" label
+        (pp_sep_list ~sep:", " (fun fmt (field, expr) ->
+             fprintf fmt "%s : %a" field pp_expr expr))
+        fields
+  | Transition { lhs; rhs; short_circuit } ->
+      fprintf fmt "%a -> %a%a" pp_expr lhs pp_expr rhs pp_short_circuit
+        short_circuit
+  | Indexed { index; list; body } ->
+      fprintf fmt "%s(%s, %s: %a)" (tok_str INDEX) index list pp_expr body
+
+and pp_application_lhs fmt =
+  let open Rule in
+  function
+  | EmptyApplicator -> ()
+  | Relation name -> pp_print_string fmt name
+  | TupleLabel name -> pp_print_string fmt name
+  | ExprOperator op -> pp_print_string fmt (expr_operator_to_string op)
+  | Fields path -> fprintf fmt "%s" (String.concat "." path)
+  | Unresolved expr -> pp_expr fmt expr
```

Comment:
What's an example of an `Unresolved`? Is this something that can't be found in `asl.spec`, e.g. when you have `asl.spec` rules referring to something manually defined?